### ISSUE TITLE
# New Feature: Purchase Order

### ISSUE THREAD

**Issue by beinafuushop** [jekkos/opensourcepos#1479]
Hello Team,
 Can we update our system by adding Purchasing Order functionality? This will give a room for creating purchasing order and send it to approver for confirmation of order. Once order delivered, In the received It should be possible to recall the order and edit the receivings if needed i.e. quantity and items. Or can receive order at it was issued.Reports for the order also needed. What do you think about this?

**Comment by RamkrishnaMondal** [comment]
First of all, we need a full functional 'Purchases' module replacing this almost useless 'Receivings' and for that we need contributor who is willing to do that.

**Comment by amrshah** [comment]
Agreed @RamkrishnaMondal

**Comment by SteveIreland** [comment]
I think it's the intent of the system architecture deciders here to try to keep OSPOS focused on sales type stuff and leave purchasing and receivables to other applications such as FrontAccounting.  

If that's changing then I need to known now before I commit too much more effort into FrontAccounting integration.

**Comment by beinafuushop** [comment]
thanx team, so what is the wY foward for this?

**Comment by amrshah** [comment]
Agreed @SteveIreland .

Is there any document outlining architectural goals?

**Comment by SteveIreland** [comment]
I shall leave those questions to @jekkos and @daN4cat  😈

**Comment by daN4cat** [comment]
Architecture goal? Here a taste....

1. do not build monoliths, we would be against any nowadays industry trend
2. move to a proper domain model oriented design at higher level which means don't mixup till functionality with back office one
3. make OSPOS modular in vertical slices so we have proper functional decomposition
4. reorganise the layering moving to a thicker client and a web middleware service channels (e.g. web, mobile, BI) via REST APIs
5. reorganise the data layer separating current data from historical data, first one follows a domain model data design, second one a dimensional one so to speed up and facilitate insight / reporting
6. ....

**Comment by daN4cat** [comment]
According to 2, OSPOS should not become an ERP tool.
Ideally one day we could have till functionality in one place and stock management in another, of course with shared data.

**Comment by SteveIreland** [comment]
I'm particularly fond of 5. :-)

**Comment by daN4cat** [comment]
Yes, but probably for the dimensional one we would need to use views so we can change the type of aggregations without impacting tables? More to ponder about even considering the efficiency (memory + latency) impact.

**Comment by beinafuushop** [comment]
Hello Team,

Am not a developer, But i can help to do documentation of the requirement or solution if needed,

Please let me know.

**Comment by daN4cat** [comment]
@beinafuushop that would be great. Please use the wiki and start writing down your guide. The rest of the group will help. Thanks

**Comment by RamkrishnaMondal** [comment]
For indian users, we need a 'Purchases' module. We also need new db table  'ospos_purchases_items_taxes' for calculation of GST paid on purchased items during a certain period. 

Will be thankful, if you consider this.

**Comment by daN4cat** [comment]
@SteveIreland I can see that you want to focus on PO only, not the full management from PO, to receiving / invoice and subsequent supplier payment. Is that intentional?

**Comment by daN4cat** [comment]
In our case we receive order confirmations from suppliers in pdf documents or even scan of manually filled papers, and then after the goods are received we receive pdf copies of invoices, or scan of invoices.
We just need to attach those files and track everything down to the payment, and also for the quarterly VAT payment (so a quarter needs to be tagged for filtering purposes).

**Comment by SteveIreland** [comment]
@daN4cat Nope.  I just haven't had time to revisit it.  I went ahead and started the page so that if anyone else wanted to add any thoughts to it then they could do so.  I definitely need to and plan to integrate receiving and invoice reconciliation.

**Comment by SteveIreland** [comment]
@RamkrishnaMondal I'll need to do a bit more research to understand the GST requirements but I'll make sure that's included.

**Comment by RamkrishnaMondal** [comment]
@SteveIreland Thanks in advance. 
Please let me know, if you need any help from me regarding the Indian GST.

**Comment by beinafuushop** [comment]
Hello Team,
I have updated the Wiki with more requirement, Can some one check and let me know if they are clear?

**Comment by beinafuushop** [comment]
@RamkrishnaMondal, @SteveIreland , @daN4cat, Can you please let us know the way foward for this.

**Comment by SteveIreland** [comment]
If you are asking if those requirements can be incorporated in the design and development then yes.  I don't personally have a need for it yet.  Our shop (at this time) is only 1 or 2 sales/office staff and the owner doesn't want to touch the computer.  So operationally there are not a lot of checks and balances, it is mostly "after the fact" reviews and reports.

I'd be interested to know the size of the companies that take advantage of OSPOS.

I expect to start the development this weekend - as soon as I finish a rewrite of the Work Order feature that we've been using in our shop for a few months.  It will target the 3.2 release.  How fast it is delivered will depend, as always, on available time.  Since it will need to play well with others I'm sure the testing will an interesting challenge.

**Comment by daN4cat** [comment]
BTW, how is this different from #566 ?

**Comment by SteveIreland** [comment]
Not a lot different.

I don't see anyone adding new requirements to the wiki so I'm assuming it's reasonably stable at the moment.  I'm going to proceed with working on it.

**Comment by SteveIreland** [comment]
Unless someone has a different thought I'm reserving the truck.png and renaming it as purchasing.png to be the icon for purchasing.  My thought was "having stuff delivered".  Since I don't expect OSPOS to ever be doing distribution management I thought it would be a safe use of the icon.

**Comment by daN4cat** [comment]
Icon is fine and I added my comments.

However if I get right this is purely purchase order and supplier confirmation up to receiving point.
There is no supplier's invoice life cycle management, which means get the invoice, set payment terms and conditions, close with a payment.

I see the purchasing as a complete story, from procurement to finance.

**Comment by SteveIreland** [comment]
What is Agency in relationship to Supplier and what role (if any) does the agency play purchasing?  I assume it's just for purposes of generating custom reports but I'm throwing it out here just to confirm.

**Comment by daN4cat** [comment]
In the case of my wife business you could buy goods directly from a company, through a distributor or agent.
The difference between distributor and agent is that the former buys and resells goods, the latter is just a salesman getting a commission on the sale.
Agents send order confirmation but do not invoice,  distributors issue order confirmation and invoice like a company does.
In short it's just to add more information to contact person as it could be an agency and not the company directly.
It's useful for reporting purposes because you want to know how much you spend by agency and play economy of scale with them to get a discount.

**Comment by SteveIreland** [comment]
I just want to confirm an assumption of mine.   Items that are sold by serial number are not received by serial number.  So my assumption is that even though the serial number is tracked at the point of sale, items do not need to be received by serial number and definitely do not need to be purchased by serial number.  Is there a scenario where this assumption wouldn't be correct?

**Comment by daN4cat** [comment]
We don't buy by serial number (barcode), it has some other sku code at purchasing time. When the items are produced they are assigned a Barcode and when they arrive in the shop they are created with the barcode. It happened in the past that we received csv files with all the list of items, barcodes and etc., and so we imported the everything using that means.

**Comment by daN4cat** [comment]
I hope you meant for serial number barcode.

**Comment by SteveIreland** [comment]
I didn't think it was related to barcode functionality.  I'm referring to the 30 character `serialnumber` field on the `sales_items` table.  I'm guessing that it is used to track the serial numbers of high value items with a warranty on them that the vendor or supplier requires to be tracked by the retailer.

**Comment by daN4cat** [comment]
Ah ok, that's used indeed for cases where each piece of a kind needs to be singularly tracked.
Indeed I don't think you can order using that field.

**Comment by scottie01** [comment]
Is it normal that customers and not suppliers are searchable/selectable when creating a purchase order? I don't seem to be able to select suppliers at all from the sales module screen. Should I be creating them elsewhere or having my suppliers duplicated in my customer list? It's probably my mistake?

**Comment by daN4cat** [comment]
Unless mistaken purchase order is not implemented, therefore I'm not sure what you are after there.

**Comment by scottie01** [comment]
Thanks DaN4cat, I'm very new to Git and hadn't looked through the wiki well enough. I was trying to use work order to generate a purchase order. My Apologies. Thanks so much to all of you for such a cool app. In my business purchase orders are critical control docs. I can see now that I will be far better off if I use work orders too. I feel very embarrassed.

**Comment by daN4cat** [comment]
Work order is something that involves you as active producer of something (a service or selling of goods).

Purchase order is a buying phase.

Therefore the two are different things.

**Comment by scottie01** [comment]
I read the wiki now. I think I've got it now. Thanks for taking the time for explanation. 

So in my business I'm a Tyre shop. We'd use work orders to issue to our staff to do work (Like a job sheet) so they're clear on the customers expectations which we've entered into the work order. Once the work is complete, the staff member can add their charges (we have standard charges saved as Item Kits) and complete the sale issuing a gst receipt and tax invoice. If I've got this correct, this is wonderful??

Once the new (yet to be released purchase order) is available would we use these to record and initiate tyre orders from our supplier so they can be reconciled against our order once received/partial receipt into stock, auto-updating our stock levels accordingly and giving us a live view of our upcoming payables and values for stock. This will be very very nice.  Have I understood correctly?

**Comment by SteveIreland** [comment]
You are correct on the use of the work orders.  I set that up a client to use in an auto repair shop.  Unfortunately for me they partnered up with a larger organization that required that they use that organization's software, so they are no longer a client of mine.  Which means that the purchase order module is no longer on my radar to develop anytime soon.  

However, if you find that you need some improvements to the work order system just post an issue here and I'll try to accommodate.

**Comment by scottie01** [comment]
Hi SteveIreland, Thanks for your contact. Yes, I'm very keen to be able to use the purchase order module. I'm currently using the work orders as purchase orders incorrectly on purpose by adding key suppliers to my customer list from the sales screen when creating the work order to be used as a purchase order. I renamed the work order to work/purchase order. I currently need to enter negative quantity too which is pretty terrible abuse your great work but it does correctly update the stock levels and creates a printed purchase order with correct details. After reading your wiki I can see many advantages in your properly thought through solution over my workaround. I think purchase orders are criticle control documents for anyone with staff and I do really like your proposed approval requirement which could've saved me many times already from unauthorized staff purchases. 
I also need to gain a far better understanding of Git, Wiki, etc sorry.

**Comment by conrado32g** [comment]
I've just finished reading the Purchase Order Wiki and the related opened issues. I need to use purchase orders, too, and have given some thought to it. The wiki shows a good start, with additional features that I had not considered. I'm sorry if this post is too long, maybe I will split parts of it off later into separate issues. (?) 

This post is to bump this issue, and indicate that I am planning to implement a purchase order module if someone else hasn't already (I couldn't find it anywhere?). I will need to learn to build the project first, but it looks like it is well documented enough that I will be able to do that. 

I will focus on my own use case; **I operate a tech service** (especially appliance repair and lock-smith services). Then, when I have achieved something functional, I will post back here in case that someone will want to improve my code (or possibly add it to the repository) and expand it to the use cases where I am not an expert (I have a lot of experience with hardware store retail and tech service applications; but not with expiration dates or serial numbers for example).

I have a few observations related to the purchase order:

- **Item codes**: Almost all of my suppliers use a unique internal SKU (Stock Keeping Unit) numeric or alpha-numeric code that is not necessarily the item's UPC, which is, well, universal rather than internal. I also use an internal SKU, which is not related to my supplier's SKU code at all, but is my own unique code. Neither of these is the EAN barcode, which is an international standard, nor is it necessarily the MPN (Manufacturer Part Number), except in the cases where I purchase directly from the manufacturer. 
- **Brand** and **MPN** are necessary information for my purchase orders because some distributors change their items for a different brand, for example a 6" pair of pliers may be a different model number next year due to cost considerations or changes from the manufacturer. These two could possibly be concatenated to the product name field as text in the requisition 'Description' field, to avoid extra fields in the 'requisition' table.
- **Items table modifications**: In my own OSPOS installation I have added the following fields to the items table: 'sku', 'brand', 'mpn', and 'provsku' (provider sku), and I have updated the Items and Sales modules to search and suggest items by these fields also, since many of my products come with the supplier's sku label attatched and/or an MPN already marked on them. The provider sku field maps directly to the "Supplier Code" in the requisition table mentioned at the wiki
- I think that a **serial number** would not be included at all in a purchase order since it is for our purposes random, being assigned at the point of manufacture or certification to each individual product.
- I added **Supplier Salesperson** field to the Purchase Order table in the wiki, since some of my suppliers have multiple sales people, so that the supplier/distributor back office (finance, dispatch, etc.) can know who to consult internally about my purchase order.
- **RFQ** option: I would like an option to print the purchase order without pricing before approval, on a RFQ (Request For Quotation) or RFB (Request For Bid) template.
Please comment if I am making an important mistake somewhere, and I will try to adjust my course.
Cheers! --Conrad

**Comment by SteveIreland** [comment]
Hi @conrado32g.  This is a really old old ticket that you've resurrected.   I don't think anyone ever committed to working on it since it would consume so much scarce time.  We were mostly playing with ideas.  Since everyone is engaged with other jobs, you're probably on your own with this.  If you fork OSPOS and make it open source I'd like to keep an eye on your effort.
